### ISSUE TITLE
2023腾讯犀牛鸟开源人才培养计划-NCNN

### ISSUE THREAD

**Issue by tencent-adm** [Tencent/OpenSourceTalent#40]
**腾讯犀牛鸟开源人才培养计划**
Tencent Rhino-bird Open-source Training Program 2021年腾讯首次发布“犀牛鸟开源人才培养计划”，通过打造面向高校学生的开源课程和开源实践培养方案，培育开源人才、普及开源文化，助力开源人才生态的发展。由教育部计算机类专业教学指导委员会、教育部软件工程专业教学指导委员会、教育部电子信息类专业教学指导委员会、开放原子开源基金会、中国信息通信研究院作为联合指导单位，腾讯开源和腾讯高校合作联合主办，培养计划涉及开源基础课程、开源进阶研学和开源项目实战三大阶段培养。
2023年将持续聚焦行业技术热点，提供企业真实项目和开发场景，提供前端、AI、大数据、基础平台、数据库等5大核心技术领域的明星项目和开源导师。
各位同学报名即可参加线上开源系列课程，学习开源知识，对话学界及产业界大咖。其中入选开源项目实战的学生将在导师的一对一指导下，开展编程实战，并有机会获 得腾讯开源贡献者证书（目前全球只发出了30+张）。 更多项目信息请看：[https://opensource.tencent.com//summer-of-code](https://opensource.tencent.com/summer-of-code)

**项目简介**
一个为手机端极致优化的高性能神经网络前向计算框架。基于 ncnn，开发者能够将深度学习算法轻松移植到手机端高效执行，开发出人工智能 APP，将 AI 带到你的指尖。

**项目导师**
倪辉，ncnn作者

**导师寄语**
ncnn是业界知名的AI推理框架，社区支持很好，参与研发是能学到知识和技术的，导师会很耐心帮助你的，对吧对吧！

**编程任务**
1(easy) in-house math routines

Some embedded platforms may not provide math library by default.
For maximum portability, it is beneficial to add in-house math routines, like simpleomp simpleocv simplestl did the similar jobs

goals:
implement all math functions used in ncnn without math.h, eg exp log sin abs etc
add cmake option NCNN_SIMPLEM for enabling
add continuation integration

2(easy) pnnx python wrapper

Though we have prebuild pnnx binary package, it is still not convenient enough to convert pytorch module into pnnx directly in python programming.
Passing input tensor shapes as command line arguments are error-prone.

goals:
create python wrapper for invoking command line pnnx executable with arguments
allow exporting torch module in python with pnnx(net, (a, b), "out.pnnx")
add continuation integration for releasing pypi package
pip install pnnx on windows/linux/macos

3(medium) optimize loongarch gemm

write gemm_loongarch.cpp with lsx extension
test with qemu

goals:
support any packed layout
good benchmark result

4(medium-hard) optimize risc-v gemm

write gemm_riscv.cpp with risc-v vector extension and zfh(fp16)
test with qemu

goals:
support any packed layout and fp32/fp16
good benchmark result

5(medium-hard) optimize vulkan softmax with shared memory and subgroup operation

The reduction part in softmax vulkan implementation is quite naive, no use of any shared memory or subgroup operation.
A decent speedup could be expected if we apply these technologies.
As softmax is a major time consuming guy inside attention op, we could get much faster stable-diffusion vulkan.

goals:
implement reduction in softmax vulkan with shared memory and subgroup operation, for all axis and data types
test speed on desktop gpu and mobile phone gpu

6(insanely hard) large language model with ncnn

It seems to be a paper task, but if you want to be the conqueror, we could help  ;)
make LLM trace-able and export to torchscript, convert with pnnx, optimize attention fusing ....
experienced cpp programming skill is essential

goals:
implement one popular LLM inference with ncnn, eg llama, chatglm or rwkv
implement interactive command line application

**联系导师**
倪辉:nihuini@tencent.com

**Comment by suzhenheng666** [comment]
issue

**Comment by Ran00w** [comment]
已成功领取本issue
